File: packages/expo-media-viewer/android/src/main/java/com/juanrdbo/mediaviewer/MediaViewerThumbnailAnchor.kt
package com.juanrdbo.mediaviewer

import org.json.JSONObject

data class MediaViewerThumbnailAnchor(
    val cornerRadiusDp: Float?,
    val contentFit: String,
) {
    fun cornerRadiusPx(density: Float): Float? = cornerRadiusDp?.let { it * density }

    companion object {
        const val CONTENT_FIT_COVER = "cover"
        const val CONTENT_FIT_CONTAIN = "contain"

        fun parse(anchorJson: String?): MediaViewerThumbnailAnchor? {
            if (anchorJson.isNullOrBlank()) return null

            val anchor = runCatching { JSONObject(anchorJson) }.getOrNull() ?: return null
            val contentFit =
                anchor
                    .optString("contentFit")
                    .takeIf { it == CONTENT_FIT_CONTAIN || it == CONTENT_FIT_COVER }
                    ?: CONTENT_FIT_COVER
            val cornerRadiusDp =
                anchor
                    .optFiniteDouble("cornerRadius")
                    ?.takeIf { it >= 0.0 }
                    ?.toFloat()

            return MediaViewerThumbnailAnchor(
                cornerRadiusDp = cornerRadiusDp,
                contentFit = contentFit,
            )
        }
    }
}

private fun JSONObject.optFiniteDouble(name: String): Double? {
    if (!has(name) || isNull(name)) return null
    val value = optDouble(name, Double.NaN)
    return value.takeIf { it.isFinite() }
}

File: packages/expo-media-viewer/ios/MediaViewerThumbnailAnchor.swift
import UIKit

struct MediaViewerThumbnailAnchor {
  static let contentFitCover = "cover"
  static let contentFitContain = "contain"

  let cornerRadius: CGFloat?
  let contentFit: String

  var contentMode: UIView.ContentMode {
    contentFit == Self.contentFitContain ? .scaleAspectFit : .scaleAspectFill
  }

  static func decode(_ json: String?) -> MediaViewerThumbnailAnchor? {
    guard let json, let data = json.data(using: .utf8) else { return nil }
    guard let payload = try? JSONDecoder().decode(DecodedThumbnailAnchor.self, from: data) else {
      return nil
    }

    let decodedContentFit = payload.contentFit
    let contentFit =
      decodedContentFit == contentFitContain || decodedContentFit == contentFitCover
        ? decodedContentFit ?? contentFitCover
        : contentFitCover
    let cornerRadius =
      payload.cornerRadius.flatMap { value -> CGFloat? in
        guard value.isFinite, value >= 0 else { return nil }
        return CGFloat(value)
      }

    return MediaViewerThumbnailAnchor(cornerRadius: cornerRadius, contentFit: contentFit)
  }
}

private struct DecodedThumbnailAnchor: Decodable {
  let cornerRadius: Double?
  let contentFit: String?
}
